mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp

```diff
@@ -0,0 +1,491 @@
+}
+
+extern "C" void mgpuStreamSynchronize(StreamWrapper *stream) {
+  if (stream)
+    stream->sync();
+}
+
+extern "C" void mgpuStreamDestroy(StreamWrapper *stream) {
+  if (stream)
+    delete stream;
+}
+
+extern "C" void mgpuStreamWaitEvent(StreamWrapper *stream,
+                                    ze_event_handle_t event) {
+  assert(stream && event);
+  stream->sync(event);
+}
+
+extern "C" ze_event_handle_t mgpuEventCreate() {
+  return getDynamicEventPool().takeEvent();
+}
+
+extern "C" void mgpuEventDestroy(ze_event_handle_t event) {
+  return getDynamicEventPool().releaseEvent(event);
+}
+
+extern "C" void mgpuEventSynchronize(ze_event_handle_t event) {
+  L0_SAFE_CALL(
+      zeEventHostSynchronize(event, std::numeric_limits<uint64_t>::max()));
+  L0_SAFE_CALL(zeEventHostReset(event));
+}
+
+extern "C" void mgpuEventRecord(ze_event_handle_t event,
+                                StreamWrapper *stream) {
+  L0_SAFE_CALL(
+      zeCommandListAppendSignalEvent(getRtContext().immCmdListCopy, event));
+  L0_SAFE_CALL(
+      zeCommandListAppendSignalEvent(getRtContext().immCmdListCompute, event));
+}
+
+extern "C" void *mgpuMemAlloc(uint64_t size, StreamWrapper *stream,
+                              bool isShared) {
+  return catchAll([&]() {
+    void *memPtr = nullptr;
+    constexpr size_t alignment{64};
+    ze_device_mem_alloc_desc_t deviceDesc = {};
+    deviceDesc.stype = ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC;
+    if (isShared) {
+      ze_host_mem_alloc_desc_t hostDesc = {};
+      hostDesc.stype = ZE_STRUCTURE_TYPE_HOST_MEM_ALLOC_DESC;
+      L0_SAFE_CALL(zeMemAllocShared(getRtContext().context, &deviceDesc,
+                                    &hostDesc, size, alignment,
+                                    getRtContext().device, &memPtr));
+    } else {
+      L0_SAFE_CALL(zeMemAllocDevice(getRtContext().context, &deviceDesc, size,
+                                    alignment, getRtContext().device, &memPtr));
+    }
+    if (!memPtr)
+      throw std::runtime_error("mem allocation failed!");
+    return memPtr;
+  });
+}
+
+extern "C" void mgpuMemFree(void *ptr, StreamWrapper *stream) {
+  stream->sync();
+  if (ptr)
+    L0_SAFE_CALL(zeMemFree(getRtContext().context, ptr));
+}
+
+extern "C" void mgpuMemcpy(void *dst, void *src, size_t sizeBytes,
+                           StreamWrapper *stream) {
+  stream->enqueueOp([&](ze_event_handle_t newEvent, uint32_t numWaitEvents,
+                        ze_event_handle_t *waitEvents) {
+    L0_SAFE_CALL(zeCommandListAppendMemoryCopy(getRtContext().immCmdListCopy,
+                                               dst, src, sizeBytes, newEvent,
+                                               numWaitEvents, waitEvents));
+  });
+}
+
+template <typename PATTERN_TYPE>
+void mgpuMemset(void *dst, PATTERN_TYPE value, size_t count,
+                StreamWrapper *stream) {
+  auto listType =
+      getRtContext().copyEngineMaxMemoryFillPatternSize >= sizeof(PATTERN_TYPE)
```

Comment:
call `getRtContext()` once and take the reference.

Reply:
Done.